### PR TITLE
[java][bidi]: add `getData`, `addDataCollector` and `removeDataCollector` commands

### PR DESCRIPTION
**Pull request by navin772**
### 💥 What does this PR do?

Adds `getData`, `addDataCollector` and `removeDataCollector` commands as in BiDi spec:

1. [w3c.github.io/webdriver-bidi#command-network-addDataCollector](https://w3c.github.io/webdriver-bidi/#command-network-addDataCollector)
2. [w3c.github.io/webdriver-bidi#command-network-getData](https://w3c.github.io/webdriver-bidi/#command-network-getData)
3. [w3c.github.io/webdriver-bidi#command-network-removeDataCollector](https://w3c.github.io/webdriver-bidi/#command-network-removeDataCollector)

### 🔄 Types of changes

- New feature (non-breaking change which adds functionality *and tests!*)

___

### **PR Type**
Enhancement

___

### **Description**
- Add BiDi network data collection commands per W3C spec

- Implement `addDataCollector`, `removeDataCollector`, and `getData` methods

- Create parameter classes for data collection configuration

- Add comprehensive test coverage for all new functionality

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Network Module"] --> B["addDataCollector"]
  A --> C["removeDataCollector"]
  A --> D["getData"]
  B --> E["AddDataCollectorParameters"]
  D --> F["GetDataParameters"]
  D --> G["BytesValue"]
  E --> H["DataType enum"]
  F --> H
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Network.java</strong><dd><code>Add network data collection methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/src/org/openqa/selenium/bidi/module/Network.java

<ul><li>Add three new methods: <code>addDataCollector</code>, <code>removeDataCollector</code>, <code>getData</code><br> <li> Import new parameter classes and JSON handling utilities<br> <li> Implement result mapping for <code>getData</code> command response</ul>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16336/files#diff-e742aac674b88cc9d9b7d49a9adf66b4dd95c479327e712cf08a3f8c02a43e7f">+44/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>AddDataCollectorParameters.java</strong><dd><code>Create AddDataCollectorParameters class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/src/org/openqa/selenium/bidi/network/AddDataCollectorParameters.java

<ul><li>New parameter class for <code>addDataCollector</code> command<br> <li> Support for data types, max size, collector type, contexts<br> <li> Builder pattern with fluent API and validation</ul>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16336/files#diff-04b07a8a086670f0cd8472a2e7f965c8b32e1f80b3373b877517301cc72b4937">+78/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>DataType.java</strong><dd><code>Add DataType enum</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/src/org/openqa/selenium/bidi/network/DataType.java

<ul><li>New enum defining supported data types<br> <li> Currently supports RESPONSE data type</ul>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16336/files#diff-32fe80ddcb406de447975b722c7ae91cf4659c9dd8c426a0fdd84ee0927c3fec">+33/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>GetDataParameters.java</strong><dd><code>Create GetDataParameters class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/src/org/openqa/selenium/bidi/network/GetDataParameters.java

<ul><li>New parameter class for <code>getData</code> command<br> <li> Support for data type, request ID, collector, disown flag<br> <li> Builder pattern with validation</ul>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16336/files#diff-892b7d00caebc07bb995b69484405ab4032b1545beedd3fb400e8575fb16e36b">+59/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>NetworkCommandsTest.java</strong><dd><code>Add data collection command tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

java/test/org/openqa/selenium/bidi/network/NetworkCommandsTest.java

<ul><li>Add comprehensive tests for all new data collection methods<br> <li> Test various parameter combinations and edge cases<br> <li> Verify error handling for invalid scenarios</ul>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16336/files#diff-56334c02fd23ebd6590a6ae6f44b2113bdbc325175c1273c4ef17a236119e517">+157/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Review by asolntsev — APPROVED** [review]

File: java/test/org/openqa/selenium/bidi/network/NetworkCommandsTest.java [hunk 1]
```diff
@@ -353,4 +354,160 @@ void throwsExceptionForInvalidContext() {
...
+      AddDataCollectorParameters parameters =
+          new AddDataCollectorParameters(List.of(DataType.RESPONSE), 1024);
+
+      String collector = network.addDataCollector(parameters);
+      assertThat(collector).isNotNull();
+    }
+  }
+
+  @Test
+  @NeedsFreshDriver
```
Comment (diemol) [inline java/test/org/openqa/selenium/bidi/network/NetworkCommandsTest.java:376]: Why can't the test reuse the driver? Is this a BiDi limitation?

Comment (navin772) [reply java/test/org/openqa/selenium/bidi/network/NetworkCommandsTest.java:376]: I added it since other tests were using it too, I will check if its strictly required or not.

Comment (navin772) [reply java/test/org/openqa/selenium/bidi/network/NetworkCommandsTest.java:376]: @diemol 3 tests fail when `@NeedsFreshDriver` is removed from all the tests, 2 auth ones and 1 data collector that was added.
Should I remove it for other tests? It does reduce the test execution time by a lot (45sec -> 14sec)

Comment (titusfortner) [reply java/test/org/openqa/selenium/bidi/network/NetworkCommandsTest.java:376]: We need to be able to clear handlers to reuse drivers

**Comment by titusfortner** [comment]
So, data collection doesn't make sense outside the context of an interception. You have to pass a request id to get the data, which is only accessible from a handler.

I think this should all be private API. When you register a handler you specify if you need the body, and that method registers the collector and manages it in the background.

**Comment by navin772** [comment]
> I think this should all be private API. When you register a handler you specify if you need the body, and that method registers the collector and manages it in the background.

@titusfortner should we overload the event handlers to something like this and manage the collector inside the event handler?:

```java
public void onBeforeRequestSent(Consumer<BeforeRequestSent> consumer, boolean captureBody)
```

Also, will this be applicable to only these events:
1. `onBeforeRequestSent`
2. `onResponseCompleted`
3. `onResponseStarted`

**Comment by newsgrep** [comment]
@pujagani @diemol @titusfortner @AutomatedTester what needs to happen to keep this MR going?
It would be a great help to get the boddy of the response/request in Java through the BiDi-API

**Comment by diemol** [comment]
We are moving to a different way of generating this code, so we're deciding soon if this one is going to get merged or if we're going to replace everything with a new approach.
